src/main/scala/is/hail/distributedmatrix/RowMatrix.scala

```diff
@@ -105,4 +161,96 @@ class RowMatrix(val hc: HailContext,
       }.filter(_.nonEmpty)
     }.writeTable(path, hc.tmpDir, header, exportType)
   }
-}
\ No newline at end of file
+}
+
+// [`start`, `end`) is the row range of partition
+case class ReadBlocksAsRowsRDDPartition(index: Int, start: Long, end: Long) extends Partition
+
+class ReadBlocksAsRowsRDD(path: String,
+  sc: SparkContext,
+  partitionCounts: Array[Long],
+  gp: GridPartitioner) extends RDD[(Long, Array[Double])](sc, Nil) {
+  
+  private val partitionStarts = partitionCounts.scanLeft(0L)(_ + _)
+  
+  if (partitionStarts.last != gp.nRows)
+    fatal(s"Error reading BlockMatrix as RowMatrix: expected ${partitionStarts.last} rows in RowMatrix, but found ${gp.nRows} rows in BlockMatrix")
+
+  if (gp.nCols > Int.MaxValue)
+      fatal(s"Cannot read BlockMatrix with ${gp.nCols} > Int.MaxValue columns as a RowMatrix")
+  
+  private val nCols = gp.nCols.toInt
+  private val nBlockCols = gp.nBlockCols
+  private val blockSize = gp.blockSize
+
+  private val d = digitsNeeded(gp.numPartitions)
+  private val sHadoopBc = sc.broadcast(new SerializableHadoopConfiguration(sc.hadoopConfiguration))
+
+  protected def getPartitions: Array[Partition] = Array.tabulate(partitionStarts.length - 1)(pi => 
+    ReadBlocksAsRowsRDDPartition(pi, partitionStarts(pi), partitionStarts(pi + 1)))
+
+  def compute(split: Partition, context: TaskContext): Iterator[(Long, Array[Double])] = {
+    val ReadBlocksAsRowsRDDPartition(_, start, end) = split.asInstanceOf[ReadBlocksAsRowsRDDPartition]
+
+    var disArray = new Array[DataInputStream](nBlockCols)
+    val buf = new Array[Byte](blockSize << 3)
+
+    var i = start
+
+    new Iterator[(Long, Array[Double])] {
+      def hasNext: Boolean = i < end
+
+      def next(): (Long, Array[Double]) = {
+        if (i == start || i % blockSize == 0) {
+          val blockRow = (i / blockSize).toInt
+          val nRowsInBlock = gp.blockRowNRows(blockRow)
+          
+          disArray = Array.tabulate(gp.nBlockCols) { blockCol =>
+            val is = gp.coordinatesBlock(blockRow, blockCol).toString
+            assert(is.length <= d)
+            val pis = StringUtils.leftPad(is, d, "0")
+            val filename = path + "/parts/part-" + pis
+
+            val dis = new DataInputStream(sHadoopBc.value.value.unsafeReader(filename))
+
+            val nColsInBlock = gp.blockColNCols(blockCol)
+
+            assert(dis.readInt() == nRowsInBlock)
+            assert(dis.readInt() == nColsInBlock)
+            if (!dis.readBoolean())
+              fatal("BlockMatrix must be stored row major on disk in order to be read as a RowMatrix")
+
+            if (i == start) {
+              val skip = (start % blockSize).toInt * (nColsInBlock << 3)
+              assert(skip == dis.skipBytes(skip)) // FIXME: can we seek?
```

Comment:
I suspect `skipBytes` seeks if the concrete implementation _can_ seek.

The documentation on `skipBytes` leaves much to be desired:
> Makes an attempt to skip over n bytes of data from the input stream, discarding the skipped bytes. However, it may skip over some smaller number of bytes, possibly zero. This may result from any of a number of conditions; reaching end of file before n bytes have been skipped is only one possibility. This method never throws an EOFException. The actual number of bytes skipped is returned.

~ [DataInput.skipBytes](https://docs.oracle.com/javase/8/docs/api/java/io/DataInput.html#skipBytes-int-)

It seems that this assertion may fail even on a properly formatted matrix file. I'm not sure if this is a practical concern. Have you run this code on a cluster against a GCP bucket?